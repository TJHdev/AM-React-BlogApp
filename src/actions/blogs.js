import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addBlog = (blog) => ({
  type: 'ADD_BLOG',
  blog: blog
})

export const startAddBlog = (blogData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      lastEditedAt = 0
    } = blogData;
    
    const blog = { title, body, createdAt, lastEditedAt };

    return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
      dispatch(addBlog({
        id: ref.key,
        ...blog
      }));
    });
  };
};

export const removeBlog = ({ id } = {}) => ({
  type: 'REMOVE_BLOG',
  id: id
});

export const startRemoveBlog = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).set(null).then(() => {
      dispatch(removeBlog({ id }));
    });
  }
};

export const editBlog = (id, updates) => ({
  type: 'EDIT_BLOG',
  id: id,
  updates
});

export const startEditBlog = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
      dispatch(editBlog(id, updates));
    }); 
  };
};

export const setBlogs = (blogs) => ({
  type: 'SET_BLOGS',
  blogs: blogs
});

export const startSetBlogs = (blogsData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
      const blogs = [];
      snapshot.forEach((childSnapshot) => {
        blogs.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

export const startSetBlogsPublic = (blogsData = []) => {
  return (dispatch, getState) => {
    return database.ref('users').once('value').then((snapshot) => {
      const blogs = [];
      const val = snapshot.val();

      Object.keys(val).forEach((childSnapshot) => {
        Object.keys(val[childSnapshot].blogs).forEach((blog) => {
          blogs.push({
            id: blog,
            ...snapshot.val()[childSnapshot].blogs[blog]
          })
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

