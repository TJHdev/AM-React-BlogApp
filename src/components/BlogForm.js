import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class BlogForm extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = { // no way to do add/edit conditional without constructor
      title: props.blog ? props.blog.title : '',
      body: props.blog ? props.blog.body :  '',
      createdAt: props.blog ? moment(props.blog.createdAt) :  moment(),
      lastEditedAt: props.blog ? moment(props.blog.lastEditedAt) :  moment(),
      calendarFocused: false,
      error: ''
    };
  };
  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };
  onBodyChange = (e) => {
    const body = e.target.value;
    this.setState(() => ({ body }));
    // // alternetive implementation
    // e.persist();
    // this.setState(() => ({ e.target.value }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) { // prevents the user from clearing this value
      this.setState(() => ({ createdAt }));
    }
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.title || !this.state.body) {
      this.setState(() => ({ error: 'Please provide title and body!' }))
    } else {
      this.setState(() => ({ error: '' })) 
      this.props.onSubmit({
        title: this.state.title,
        body: this.state.body,
        createdAt: this.state.createdAt.valueOf(),
        lastEditedAt: moment().valueOf()
      }) // we pass the data up so that the form can remain re-usable
    }
  };
  render() {
    return (
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input
            type="text"
            placeholder="Title"
            className="text-input"
            autoFocus
            value={this.state.title}
            onChange={this.onTitleChange}
          />
          <textarea
            placeholder="Add the text body"
            className="textarea"
            value={this.state.body}
            onChange={this.onBodyChange}
          ></textarea>
          <div>
            <button className="button">Save Blog</button>
          </div>
        </form>
    );
  }
}