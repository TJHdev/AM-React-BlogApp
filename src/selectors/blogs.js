import moment from 'moment';

// Get visible blogs

const getVisibleBlogs = (blogs, { text, sortBy }) => {
  return blogs.filter((blog) => {
    const textMatch = blog.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'dateCreated') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'dateEdited') {
      return a.lastEditedAt < b.lastEditedAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title < b.title ? 1 : -1;
    }
  });
};

export default getVisibleBlogs;