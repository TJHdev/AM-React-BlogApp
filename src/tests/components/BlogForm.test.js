import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import BlogForm from '../../components/BlogForm';
import blogs from '../fixtures/blogs';

test('should render BlogForm correctly', () => {
  const wrapper = shallow(<BlogForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render BlogForm correctly with blog data', () => {
  const wrapper = shallow(<BlogForm blog={blogs[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid submission', () => {
  const wrapper = shallow(<BlogForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').prop('onSubmit')({
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('Please provide title and body!');
  expect(wrapper).toMatchSnapshot();
});

test('should set title on input change', () => {
  const value = 'New title';
  const wrapper = shallow(<BlogForm />);
  wrapper.find('input').prop('onChange')({
    target: { value: value }
  });
  expect(wrapper.state('title')).toBe('New title');
});

test('should set body on input change', () => {
  const value = 'New body';
  const wrapper = shallow(<BlogForm />);
  wrapper.find('textarea').prop('onChange')({
    target: { value: value }
  });
  expect(wrapper.state('body')).toBe('New body');
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<BlogForm blog={blogs[1]} onSubmit={onSubmitSpy}/>)
  wrapper.find('form').prop('onSubmit')({
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    title: blogs[1].title,
    body: blogs[1].body,
    createdAt: blogs[1].createdAt,
    lastEditedAt: moment().valueOf(),
  });
});