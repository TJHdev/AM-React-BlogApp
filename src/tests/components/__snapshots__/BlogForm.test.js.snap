// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add the text body"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Blog
    </button>
  </div>
</form>
`;

exports[`should render BlogForm correctly with blog data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value="are you sure you want to do this?"
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add the text body"
    value="Body bOdy boDy bodY"
  />
  <div>
    <button
      className="button"
    >
      Save Blog
    </button>
  </div>
</form>
`;

exports[`should render error for invalid submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add the text body"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Blog
    </button>
  </div>
</form>
`;

exports[`should render error for invalid submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add the text body"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Blog
    </button>
  </div>
</form>
`;
